datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,2001 @@
+        let join_on = vec![
+            (
+                Column::new_with_schema("c", &schema()).unwrap(),
+                Column::new_with_schema("c1", &right.schema()).unwrap(),
+            ),
+            (
+                Column::new_with_schema("b", &schema()).unwrap(),
+                Column::new_with_schema("b1", &right.schema()).unwrap(),
+            ),
+            (
+                Column::new_with_schema("a", &schema()).unwrap(),
+                Column::new_with_schema("a1", &right.schema()).unwrap(),
+            ),
+        ];
+        let bottom_right_join = ensure_distribution_and_ordering(
+            hash_join_exec(left, right.clone(), &join_on, &JoinType::Inner),
+            10,
+        );
+
+        // Join on (B == b1 and C == c and AA = a1)
+        let top_join_on = vec![
+            (
+                Column::new_with_schema("B", &bottom_left_projection.schema()).unwrap(),
+                Column::new_with_schema("b1", &bottom_right_join.schema()).unwrap(),
+            ),
+            (
+                Column::new_with_schema("C", &bottom_left_projection.schema()).unwrap(),
+                Column::new_with_schema("c", &bottom_right_join.schema()).unwrap(),
+            ),
+            (
+                Column::new_with_schema("AA", &bottom_left_projection.schema()).unwrap(),
+                Column::new_with_schema("a1", &bottom_right_join.schema()).unwrap(),
+            ),
+        ];
+
+        let join_types = vec![
+            JoinType::Inner,
+            JoinType::Left,
+            JoinType::Right,
+            JoinType::Full,
+            JoinType::LeftSemi,
+            JoinType::LeftAnti,
+            JoinType::RightSemi,
+            JoinType::RightAnti,
+        ];
+
+        for join_type in join_types {
+            let top_join = hash_join_exec(
+                bottom_left_projection.clone(),
+                bottom_right_join.clone(),
+                &top_join_on,
+                &join_type,
+            );
+            let top_join_plan =
+                format!("HashJoinExec: mode=Partitioned, join_type={:?}, on=[(Column {{ name: \"AA\", index: 1 }}, Column {{ name: \"a1\", index: 5 }}), (Column {{ name: \"B\", index: 2 }}, Column {{ name: \"b1\", index: 6 }}), (Column {{ name: \"C\", index: 3 }}, Column {{ name: \"c\", index: 2 }})]", &join_type);
+
+            let reordered = reorder_join_keys_to_inputs(top_join);
+
+            // The top joins' join key ordering is adjusted based on the children inputs.
+            let expected = &[
+                top_join_plan.as_str(),
+                "ProjectionExec: expr=[a@0 as A, a@0 as AA, b@1 as B, c@2 as C]",
+                "HashJoinExec: mode=Partitioned, join_type=Inner, on=[(Column { name: \"a\", index: 0 }, Column { name: \"a1\", index: 0 }), (Column { name: \"b\", index: 1 }, Column { name: \"b1\", index: 1 }), (Column { name: \"c\", index: 2 }, Column { name: \"c1\", index: 2 })]",
```

Comment:
I don't understand the need to reorder the join keys here -- if the input is partitioned on `a, b, c` the correct answer will still happen if you join on `b, c, a`

My rationale is that any particular values of `a`, `b` and `c` will always be the same partition if it partitioned by `hash(a, b, c)` or if it is partitioned by `hash(b, c, a)` although the specific partitions would be different it shouldn't matter

Reply:
For Aggregations, yes, the ordering doesn't matter. But for Partitioned Joins, it matters. 
For Partitioned Joins, there are two inputs, if the left is partitioned on (B, C, A) and right is partitioned on (A, C, B), for the same particular value pairs(a , b,  c), the two inputs will not co-locate to the same partition, because the hashed values are different.